Make WaylandKeyboard to notify about keyboard modifiers.

msisov opened the pull request:
This patch makes WaylandKeyboard to notify observers about
keyboard modifiers (ctrl, alt, shift and etc). Interested
clients are WaylandPointer, WaylandTouch(to be checked yet) and
maybe others. ModifiersObserver is introduced now in order to
quickly add other interested parties in the future.

In this patch WaylandPointer is the only observer now. It gets
a notification and modifies its flags with modifiers, which
are used to create MouseEvents, for example. Thus, MouseEvent
can now say if that has been a pointer event with keyboard modifiers
(alt down, ctrl down and etc) or not.

Issue #193

msisov committed:
Make WaylandKeyboard to notify about keyboard modifiers.

This patch makes WaylandKeyboard to notify observers about
keyboard modifiers (ctrl, alt, shift and etc). Interested
clients are WaylandPointer, WaylandTouch(to be checked yet) and
maybe others. ModifiersObserver is introduced now in order to
quickly add other interested parties in the future.

In this patch WaylandPointer is the only observer now. It gets
a notification and modifies its flags with modifiers, which
are used to create MouseEvents, for example. Thus, MouseEvent
can now say if that has been a pointer event with keyboard modifiers
(alt down, ctrl down and etc) or not.

Issue #193

ganguin reviewed (commented):
Thanks for your work. I applied your patch, however modifiers are still not working for me (wayland 1.13.0, weston 1.11.0 or sway 0.14.0). Is it supposed to work as is or is there another patch needed?

msisov commented:
@ganguin, it is supposed to work only with "US" layout. other layouts are not supported yet.

msisov commented:
@tonikitoo, could you check wayland_pointer.cc ? I've changed it as Rob requested and added VerifyFlagsAfterMasking. Do you think that what he has requested?

tonikitoo commented on ui/ozone/platform/wayland/wayland_pointer.h:
```diff
@@ -57,12 +57,17 @@ class WaylandPointer {
 
   void SetSerial(uint32_t serial);
 
+  int GetFlagsWithKeyboardModifiers();
```
maybe GetKeyboardModifierFlags?

tonikitoo replied:
hum, ok. maybe not. flags can contain keyboard modifiers and more flags, it seems.

msisov replied:
Right. They can contain keyboard modifiers. This method returns current |flags_| with keyboard modifiers.

tonikitoo commented on ui/ozone/platform/wayland/wayland_pointer.h:
```diff
@@ -57,12 +57,17 @@ class WaylandPointer {
   int flags_ = 0;
+
+  // Keeps track of current modifiers.
+  int modifiers_ = 0;
```
do we need modifiers_ at all? instead, maybe we can always just call connection_->GetKeyboardModifiers?

if we ultimately need it, then rename it to keyboard_modifiers_ ?

msisov replied:
We keep track of old modifiers in order to remove them before updating current flags with new modifiers. Otherwise, you never know if one must remove or add additional flags(modifiers) to current flags.
